frontend/src/metabase/query_builder/components/DimensionList.jsx

```diff
@@ -31,41 +31,51 @@ export default class DimensionList extends Component {
   };
 
   UNSAFE_componentWillMount() {
-    this._updateSections(this.props.sections);
+    this.updateSections(this.props.sections);
   }
+
   UNSAFE_componentWillReceiveProps(nextProps) {
     if (this.props.sections !== nextProps.sections) {
-      this._updateSections(nextProps.sections);
+      this.updateSections(nextProps.sections);
     }
   }
-  _updateSections(sections) {
-    this.setState({
-      sections: (sections || []).map(section => ({
-        ...section,
-        items: section.items.map(item => ({
-          ...item,
-          name: item.name || (item.dimension && item.dimension.displayName()),
-          icon: item.icon || (item.dimension && item.dimension.icon()),
-        })),
-      })),
-    });
+
+  buildUpdatedSectionItems(items) {
+    return items.map(item => ({
+      ...item,
+      name: item.name || item.dimension?.displayName(),
+      icon: item.icon || item.dimension?.icon(),
+    }));
   }
 
-  _getDimensions() {
+  updateSections(sections = []) {
+    const sectionsForNewState = sections.map(section => ({
+      ...section,
+      items: this.buildUpdatedSectionItems(section.items),
+    }));
+
+    const newState = {
+      sections: sectionsForNewState,
+    };
+
+    this.setState(newState);
+  }
+
+  getDimensions() {
     return (
       this.props.dimensions ||
       (this.props.dimension ? [this.props.dimension] : [])
     );
   }
 
   itemIsSelected = item => {
-    const dimensions = this._getDimensions();
+    const dimensions = this.getDimensions();
     const { dimension } = item;
     return (
       item.dimension &&
-      _.any(dimensions, d => {
+      _.any(dimensions, dimensionEntry => {
```

Comment:
👍 for the full name, though what do you think about make it more concise and letting it be just `dimension` (we can also remove line 73 that has the name clash as it doesn't really do anything much since we can still refer to it as `item.dimension` like line 75 anyway)?

Reply:
Yes, good call, updated.